### ISSUE TITLE
Future of Coding Weekly 2022/05 Week 5

### ISSUE THREAD

**Issue by marianoguerra** [marianoguerra/future-of-coding-weekly#134]
- [x] Change Month and Week Number
- [x] focWeekExport "2022-01-19" "2022-01-26"
- [x] Summary
- [ ] Hashtags
- [x] Check that comment links work (push weekly dump with channel summaries)
- [x] Check to mention right person for moved messages
- [x] Update Search Index
- [x] Download New Attachments
- [x] Update links

- https://tinyletter.com/
- https://tinyletter.com/marianoguerra/letters/
- http://localhost:8000/history/
- https://marianoguerra.github.io/future-of-coding-weekly/
- https://stackedit.io/app#

**Comment by marianoguerra** [comment]
🧑‍🔬 Papers We Love 💬 Thinking Together (a lot) 🔬 Interval Research Corp 📢 Unison Conference

# Our Work

💻 [henosia.com](https://www.henosia.com/) via **Jim Meyer**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-23T15:49:14.120Z)

Curious about what this group thinks of the approach we're taking with [henosia.com](https://www.henosia.com/) We provide direct manipulation of the visual output of a running program, and think this is a great fit for UX Designers where their intent is "move that", "adjust this whitespace" and "insert a button below". It's visual TypeScript code editing in real time for designers. Oh, and we have multiplayer in the mix, which has been, let's call it  _interesting_ , so far 😂

**Comment by marianoguerra** [comment]
🎥 [Banghashat demo](https://www.youtube.com/watch?v=bi3YkY7UKmM) via [Chris Knott](https://github.com/ChrisKnott)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-23T18:44:55.527Z)

![Thumbnail](https://img.youtube.com/vi/bi3YkY7UKmM/hqdefault.jpg)

I made a little demo of a text based knowledge management/second brain thingy

**Comment by marianoguerra** [comment]
🐦 [Tweet from @tobyshooters](https://twitter.com/tobyshooters/status/1529153009032253440?s=21&t=xULGxPhOKg3n6vR5UkVdJg) via **Cristóbal**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-24T18:04:40.538Z)

Tiny demo from some weekend hacking to bring the computer out of its box by mapping macOS spaces to physical spaces in my room



> 🐦 [cristóbal](https://twitter.com/@tobyshooters): Pinning macOS spaces to physical spaces: Forgot my dictionary at my desk... walk over to grab it!
> 
> Depending on one's physical location, a different macOS space is focused (see top-right). You can move windows between spaces by walking over, grabbing onto it, and walking back.
> 
> ![Tweet Thumbnail](https://pbs.twimg.com/ext_tw_video_thumb/1529151717153054722/pu/img/DD9Lp6NCWFKAx-w0.jpg)

**Comment by marianoguerra** [comment]
📢 [Papers We Love Education Edition](https://github.com/papers-we-love/edu/discussions/5) via [Joe Nash](https://twitter.com/jna_sh)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-25T11:46:17.795Z)

It's a big Papers We Love: Education edition tomorrow! We’re being joined by Jane Waite, researcher at the Raspberry Pi foundation, to talk about her report (co-authored with Sue Sentence) that summarises the research around programming pedagogy, drawing on 170 papers. All the info here: [Teaching programming in schools: A review of approaches and strategies](https://github.com/papers-we-love/edu/discussions/5)

**Comment by marianoguerra** [comment]
📢 [makeabetter.computer](http://makeabetter.computer) via **Parker Henderson**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-25T18:17:09.521Z)

Hi everyone! Sharing a project I just launched that may be of interest to some of you!



[makeabetter.computer](http://makeabetter.computer)—A (tiny) microgrants for people working and thinking about the future of computing!



I wrote a small [twitter thread](https://twitter.com/parkerhendo/status/1529502268625629184?s=20&t=Lbq7Tem7DlG09qqB3Vazdw) about it.

[🧵 another conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/linking-together.html#2022-05-23T18:19:19.138Z)

**Comment by marianoguerra** [comment]
🔌 [egui_cable: Generic and extensible data-oriented widget for connecting ports by cables](https://github.com/ryo33/egui_cable) via **Ryo Hirayama**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-26T03:50:44.534Z)

I’ve released a widget library for the visual programming editor of Desk. It's also generic and extensible widget for any purpose. You can create node-based or analog synthesizer-like UI with simple API.


![Image from iOS](http://history.futureofcoding.org/history/msg_files/F03/F03H1TFJ7FV.png)

![Image from iOS](http://history.futureofcoding.org/history/msg_files/F03/F03GZ0XTVQD.jpg)

**Comment by marianoguerra** [comment]
🤖 [Codeball -&gt; AI-powered code review](https://codeball.ai/) via **Kiril Videlov**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/share-your-work.html#2022-05-27T14:10:48.875Z)

Hey folks, I wanna show you  something new we have been working on - an  *AI code review bot* 🤖. Most pull requests (about 2/3) get merged without corrections, and the idea is to have the bot detect and approve them. This was born out frustration constantly waiting for 'code review' for small and safe code contribution, especially since I believe in "ship small & often. Also it allows you to focus on the trickier changes instead.



👉 Here's how it works: anybody can do a 'historical' test on any repository here - [codeball.ai](https://codeball.ai/) to test out the accuracy. To do reviews on new pull requests, it's setup as a GitHub action.

I'd love to hear what you smart people think! Happy to also share more about the model too (basically it's a neural net, the inputs include features about the code, the author and author experience with the code changed, trained on over 1M code contributions).

**Comment by marianoguerra** [comment]
# Thinking Together

💬 [Mariano Guerra](https://twitter.com/warianoguerra)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-23T12:28:06.510Z)

Any good recommendations on Term Rewriting Systems?

**Comment by marianoguerra** [comment]
💬 **Jim Meyer**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-24T09:01:09.317Z)

I've been thinking about the limitations and shortcomings that VPL environments have when they focus on program structure instead of the desired outcome a user has in mind.



Another word for structure is recipe. Recipes are difficult for most non-coders to wrestle with because they force a mental model of “do A, then B, then C to hopefully achieve the X that I want”. A recipe by its definition means that the outcome – the thing the user is trying to achieve - is a separate thing, which usually divides the UI into two distinct and somewhat disconnected parts.



This brings me to outcome-driven coding (ODC). I’m making up this term, surely there is a term grounded in research? 😁 With ODC, we’re flipping the script, and letting users work in a way that follows a pattern of “Tinker and change until it (the outcome, not the recipe) feels right”.



By mostly hiding the program structure, and showing the outcome of running the program live, an ODC tool can be made on top of an existing programming language, leveraging an entire eco-system of existing work, be it open source, or existing code bases that a product team is working on. This is a big deal for adoption of such a tool, and the foundation for collaboration with developers using IDEs.



There’s an interesting trade-off here that has to do with expressive power versus complexity in a VPL environment. With ODC, we can manage that complexity by deciding on which parts of a code base that visual programming is best suited for, and which parts to “leave alone but still execute to show the real product outcome”. Parts of a code base are inherently about the human element and the user experience. Let's call this "design code". Design code naturally centers on using and defining design system components, the visual language, use of data, navigation flows, and pages. This is where visual programming shines.



The ODC approach lets us create a tool that is optimized to work with design code visually, in a way that UX designers prefer, while allowing traditional business logic code to execute unhampered inside the tool. The key is to allow collaboration on a single shared code base, with tools specialized for the unique skills and needs of the people in a cross-functional team.



Would be great to hear what this group thinks of this hybrid approach, and whether there is existing work that is similar. Thanks 🙂

![image.png](http://history.futureofcoding.org/history/msg_files/F03/F03GHKG3VRU.png)

**Comment by marianoguerra** [comment]
📝 [Students who grew up with search engines might change STEM education forever](https://www.theverge.com/22684730/students-file-folder-directory-structure-education-gen-z) via [Joe Nash](https://twitter.com/jna_sh)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-27T10:51:53.043Z)

TL;DR: many computing applications have transcended files-and-folders, why haven't developer tools? 🤔



During Papers We Love Education yesterday, we were joined by a couple of teachers of various age groups, and I had the opportunity to ask them about whether they've observed what this Verge article describes: [theverge.com/22684730/students-file-folder-directory-structure-education-gen-z](https://www.theverge.com/22684730/students-file-folder-directory-structure-education-gen-z)



The article is about how ubiquitous search and cloud storage have meant that young folk no longer have a mental model of file systems, or really understand files and folders. I had suspected I was observing this in my work, with some of the questions I was getting in our introductory content about CLIs. The educators confirmed they're also seeing this and it's a big problem. 



That struck me as a big issue for CS education. If students are coming in never having interacted with an app that deals in files, saving files, or organising content in directories, that's a huge hurdle to beginning programming, or interacting with most tools. Everything we do and use still has the file system at the centre of the experience and the interface. 



Whether the “tabletification” of computing is a positive thing or not, it's interesting to me that unlike many categories of applications, developer tools, even for young learners, have resisted it

**Comment by marianoguerra** [comment]
💬 **Jim Meyer**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-27T14:50:04.176Z)

Undo and redo in relationship to CRDTs/multiplayer seems to be an underexplored area with huge challenges.



Here's a case where multiplayer undo-redo becomes interesting for visual programming.



 _(This would be visual programming, e.g. via a blocks/node interface, but I'll use JavaScript and React to express the states)_ 


[🧵 read more](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-27T14:50:04.176Z)

**Comment by marianoguerra** [comment]
💬 **Jarrett Atkinson**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-27T21:56:57.190Z)

I'm not sure if this is going to be discovery or thinking together, but oh well.



I've had the idea in the back of my mind for a long time that UI sucks. It's bad for us as programmers when we have to build an app across 5+ form factors (more when you consider the possibility of AR, VR, or voice assistant applications). It's bad for the consumer when we develop a bad UI. It's bad for the consumer when we build a good UI, but it's not talored to how individuals use it. For example I use about 5% of the potential functionality of my IDE, but it's extremely tedious ranging to impossible to customize the UI for myself. Yes, keyboard bindings exist but for the most part let's ignore that. Most casual users won't learn Keybindings. And, that ignores that I'd like to comstomize the application state I can view.



OK, the problem established, I'd like to think about a solution: stop writing UI. Instead, write api descriptors so an agent can design a UI for you, knowing the capabilities of your application. Early versions of an agent will require massive amounts of hints, but that's probably still easier than writing UI. Having an agent (ai maybe?) would also allow input from individual users to the system with hints as to how they want to view the app. "show the buttons in this order. " "show the status of this in that box." etc. Graphql is the closest approximation of such an api descriptor I can think of.



On to the questions: Has this been considered before? What prior art is in this space? Are we near technical feasibility of such a project?

**Comment by marianoguerra** [comment]
💬 **Václav Blažej**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-29T14:37:26.268Z)

 *Core of a futuristic system - use cases.*  
 A) What things should the programmer be able to do but is not able to because of the limitations of today's computers and systems? 
 B) How should a interaction of a programmer with a computer look like in the future? 
 C) If you had 10000 hours of free unencumbered development hours, what would you do?

**Comment by marianoguerra** [comment]
💬 **Evan Payne**

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/thinking-together.html#2022-05-29T19:48:17.324Z)

The details are fuzzy now, but when I was a teenager I read Tad Williams's series Otherland.  It was written in the late 90's and so didn't really grasp how the internet would come to work over time.  Its premise was that a lot of the world was immersed in virtual reality worlds, but one group of ultra rich folk had built a secret system that was impossibly realistic, and well, it got weird from there.   Still, it had some interesting concepts that really inspired me.  Two in particular:




* Sellars, and older guru-like figure, had a garden at his home on the military base which was secretly his access to his file system.  Eg, each leaf and stem on the plant was representative of files and connection between them, and they could only be accessed by being in the physical space with the plants.
* The assassin character had a generative music implant in his head, which allowed him to mentally tweak the inputs on the music playing in his head, triggering swells in the orchestra, or moving from slow to fast rhythm, or changing keys or genres.  Not playlists either, but properly generative, with inputs from his own subvocal commands, but also from the environment (eg, light or other sounds)



I think there are probably a wealth of future-of-coding ideas in older scifi novels.

**Comment by marianoguerra** [comment]
# Content

💻 [github.com/yatima-inc/yatima-lang-alpha](https://github.com/yatima-inc/yatima-lang-alpha) via [Christopher Galtenberg](https://twitter.com/galtenberg)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/linking-together.html#2022-05-23T23:20:56.397Z)

"Yatima is a pure functional programming language implemented in Rust with... content addressing, first-class types, linear, affine and erased types, type-safe dependent metaprogramming"

**Comment by marianoguerra** [comment]
📝 [Interval Research Corporation: a 90s PARC without a Xerox](https://instadeq.com/blog/posts/interval-research-corporation-a-1990s-parc-without-a-xerox/) via [Mariano Guerra](https://twitter.com/warianoguerra)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/linking-together.html#2022-05-24T10:22:12.003Z)

I came across a big research lab from the 90s that I didn't know about so I decided to blog about it:


![IntervalResearchCorporation.png](http://history.futureofcoding.org/history/msg_files/F03/F03GQLD3Y82.png)

**Comment by marianoguerra** [comment]
🐦 [Tweet from @unisonweb](https://twitter.com/unisonweb/status/1529087283378458625) via [Mariano Guerra](https://twitter.com/warianoguerra)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/linking-together.html#2022-05-24T13:18:46.092Z)

Registrations for Unison Forall 2022 are open



June 24th (EDT), fully online and completely free


> 🐦 [Unison](https://twitter.com/@unisonweb): We're very pleased to announce that registrations for Unison Forall 2022 are open! Join us June 24th (EDT) for the first Unison language conference, fully online and completely free. We have a lineup of engaging talks, and a virtual "hallway track". 🤖🌸🪐
> https://hopin.com/events/unison-forall

**Comment by marianoguerra** [comment]
📢 [Emergent Ventures Grant/Fellowship Application](https://mercatuscenter.formstack.com/forms/emergent_ventures) via [Jamie Brandon](https://scattered-thoughts.net/)

[🧵 conversation](https://marianoguerra.github.io/future-of-coding-weekly/history/weekly/2022/05/W5/linking-together.html#2022-05-24T22:45:15.822Z)

Emergent Ventures is a low-application-overhead grant for 'high-reward ideas'. I haven't been able to find out what typical grant size is, but the parallel Covid Fast Grants program seems to be averaging $250k.

[Previous grants](https://docs.google.com/spreadsheets/d/1NYv0vsWVtEMDvO7gYDh4DDH4oNBGQBGpV0bZzG9x9JY/edit#gid=243061127)

**Comment by marianoguerra** [comment]
https://tinyletter.com/marianoguerra/letters/future-of-coding-weekly-2022-05-week-5
